cdap-app-fabric/src/main/java/org/apache/twill/internal/yarn/AbstractYarnTwillService.java

```diff
@@ -0,0 +1,159 @@
+  protected volatile Credentials credentials;
+
+  protected AbstractYarnTwillService(ZKClient zkClient, RunId runId, Location applicationLocation) {
+    super(zkClient, runId);
+    this.applicationLocation = applicationLocation;
+  }
+
+  /**
+   * Returns the location of the secure store, or {@code null} if either not running in secure mode or an error
+   * occur when trying to acquire the location.
+   */
+  protected final Location getSecureStoreLocation() {
+    if (!UserGroupInformation.isSecurityEnabled()) {
+      return null;
+    }
+    try {
+      return applicationLocation.append(Constants.Files.CREDENTIALS);
+    } catch (IOException e) {
+      LOG.error("Failed to create secure store location.", e);
+      return null;
+    }
+  }
+
+  /**
+   * Attempts to handle secure store update.
+   *
+   * @param message The message received
+   * @return {@code true} if the message requests for secure store update, {@code false} otherwise.
+   */
+  protected final boolean handleSecureStoreUpdate(Message message) {
+    if (!SystemMessages.SECURE_STORE_UPDATED.equals(message)) {
+      return false;
+    }
+
+    // If not in secure mode, simply ignore the message.
+    if (!UserGroupInformation.isSecurityEnabled()) {
+      return true;
+    }
+
+    try {
+      Credentials credentials = new Credentials();
+      Location location = getSecureStoreLocation();
+      try (DataInputStream input = new DataInputStream(new BufferedInputStream(location.getInputStream()))) {
+        credentials.readTokenStorageStream(input);
+      }
+
+      UserGroupInformation.getCurrentUser().addCredentials(credentials);
+
+      // CDAP-5844 Workaround for HDFS-9276, to update HDFS delegation token for long running application in HA mode
+      cloneHaNnCredentials(location, UserGroupInformation.getCurrentUser());
+      this.credentials = credentials;
+
+      LOG.info("Secure store updated from {}.", location);
+
+    } catch (Throwable t) {
+      LOG.error("Failed to update secure store.", t);
+    }
+
+    return true;
+  }
+
+  private static void cloneHaNnCredentials(Location location, UserGroupInformation ugi) throws IOException {
+    Configuration hConf = getConfiguration(location.getLocationFactory());
+    String scheme = location.toURI().getScheme();
+
+    Map<String, Map<String, InetSocketAddress>> nsIdMap = DFSUtil.getHaNnRpcAddresses(hConf);
+    for (Map.Entry<String, Map<String, InetSocketAddress>> entry : nsIdMap.entrySet()) {
+      String nsId = entry.getKey();
+      Map<String, InetSocketAddress> addressesInNN = entry.getValue();
+      if (!HAUtil.isHAEnabled(hConf, nsId) || addressesInNN == null || addressesInNN.isEmpty()) {
+        continue;
+      }
+      // The client may have a delegation token set for the logical
+      // URI of the cluster. Clone this token to apply to each of the
+      // underlying IPC addresses so that the IPC code can find it.
+
+      URI uri = URI.create(scheme + "://" + nsId);
```

Comment:
@poornachandra you mean, in the case of multiple FileSystems?
What we could do is open a JIRA to look into that, since that's an additional dimension of complexity, and handle that separately.

@chtyim what do you think?